Verify each target has the right number of tests

cschanaj opened the pull request:
blocked by #11187 which has been pending review for a long time... Also, merge #11942 before whitelisting `rules-test`.

See https://gist.github.com/cschanaj/b3c80c935226c55e78013d81d618cd99

Related: #7272

cschanaj commented on test/rules/src/https_everywhere_checker/rules.py:
```diff
@@ -180,15 +180,55 @@ def _determineTestApplication(self):
+			if not "*" in target and not self.excludes(("http://%s/" % target)):
+				continue
+
+			# '*.example.com' match 'www.example.com' but not 'secure.account.exmple.com'
```
@Hainish does `*.example.com` matches `www.example.com`, `secure.account.example.com` and `example.com`? I am a bit confused seeing https://support.google.com/customsearch/answer/71826?hl=en

Hainish replied:
@cschanaj the logic is here: https://github.com/EFForg/https-everywhere/blob/454fb4a9a8f60ed7db6e0624637574b0534a6a14/chromium/rules.js#L350-L355

So to answer, for a.b.examble.com, *.example.com *is* a matching target.

cschanaj commented:
@Hainish I have fixed the wildcard matching logic in c255212 and this is ready for an review. thanks!!

P.S. please have a look at #11942 which fix a trivial bug in the ruleset whitelist cleanup script.

Hainish commented:
Got it, I should have some time in the next few days to review this.

cschanaj commented:
@Hainish I have extracted the matching logic from `rule.js`, it seems that if `example.*` does not match`example.co.uk` however `example.com` does, i.e. 

```xml
<target host="example.*" />
```
does not work for `example.co.uk`. It seems like a bug to me, can you verify it?

```javascript
var myfunct = function(host) {
	var tmp;
	var results = [];

	// Ensure host is well-formed (RFC 1035)
	if (host.indexOf("..") != -1 || host.length > 255) {
		log(WARN,"Malformed host passed to potentiallyApplicableRulesets: " + host);
		return null;
	}

	// Replace each portion of the domain with a * in turn
	var segmented = host.split(".");
	for (let i=0; i < segmented.length; i++) {
		let tmp = segmented[i];
		segmented[i] = "*";
		// results = results.concat(this.targets[segmented.join(".")]);
		console.log(segmented.join("."))
		segmented[i] = tmp;
	}
	// now eat away from the left, with *, so that for x.y.z.google.com we
	// check *.z.google.com and *.google.com (we did *.y.z.google.com above)
	for (var i = 2; i <= segmented.length - 2; ++i) {
		var t = "*." + segmented.slice(i,segmented.length).join(".");
		console.log(t);
		// results = results.concat(this.targets[t]);
	}
}

console.log('**** example.co.uk ****');
myfunct('example.co.uk');

console.log('**** example.com  ****');
myfunct('example.com');
```

```bash
$ nodejs rules.js 
**** example.co.uk ****
*.co.uk
example.*.uk
example.co.*
**** example.com  ****
*.com
example.*

```

cschanaj commented:
@Hainish It seems that we treat `*` in different position differently, according to my understanding of https://github.com/EFForg/https-everywhere/pull/11998#issuecomment-324275475, it seem that 

| `target` | `regex` interpretation |
| :---: | :---: |
| `*.example.com` | `.+\.example\.com$` |
| `secure.*.example.com` | `^secure\.[^\.]*\.example\.com$` |
| `www.example.*` | `^www\.example\.[^\.]*` |

cschanaj commented:
@Hainish merge #12091 will fix Travis from this PR. thanks!

cowlicks commented:
@cschanaj I'm looking into this today! Sorry for the delay.

cowlicks commented on test/rules/src/https_everywhere_checker/rules.py:
```diff
@@ -180,15 +180,69 @@ def _determineTestApplication(self):
+				continue
+
+			# According to the logic in rules.js available at
+			# EFForg/https-everywhere/blob/master/chromium/rules.js#L350-L355
```
Can you include the commit hash in this URL? This way we can find the referenced code if it moves. You can get a permalink to lines in a specific version of the file by selecting the lines normally, then pressing "y".

cowlicks commented on test/rules/src/https_everywhere_checker/rules.py:
```diff
@@ -180,15 +180,69 @@ def _determineTestApplication(self):
+				pattern = pattern.replace('*', '.+')
+
+			# however, `example.*` match `example.com` but not `example.co.uk`
+			if pattern[len(pattern) - 1] == '*':
```
You can get the last element like this: `pattern[-1]`

cowlicks commented:
@cschanaj I'm not very familiar with running these tests. Would you mind providing a testcase this PR fixes? I would just run it by pointing the `rulesdir` in `checker.config` to it, right?

cschanaj commented:
@cowlicks I run the rules test by the following command

```bash
$ export TEST=rules
$ ./test.sh
```

cowlicks commented on test/rules/src/https_everywhere_checker/rules.py:
```diff
@@ -180,15 +180,69 @@ def _determineTestApplication(self):
 			 Returns an array of strings reporting any coverage problems if they exist,
 			 or empty list if coverage is sufficient.
 			 """
 		problems = self._determineTestApplication()
```
Every "`problem`" gets printed twice, they are created here in `getCoverageProblems`. This happens because `problems` is a reference to `self.test_application_problems` through `self._determineTestApplication`.  Then again in `getTargetValidityProblems` we do `problems = self._determineTestApplication`, so the `problems` from `getCoverageProblems` also get reported through `getTargetValidityProblems`.

I'm not sure why `problems` is assigned like this. I'm guessing it is a mistake in either this function or `getTargetValidityProblems`. Maybe @Hainish can chime in.

cowlicks replied:
@cschanaj You can probably just change this to:
```javascript
self._determineTestApplication()
problems = []
```
that is what other tests do

cowlicks commented:
@cschanaj  git can be tricky sometimes. I can fix this for you if you'd like.

cowlicks commented:
I revisited this today. It looks good to me. It should be merged in conjunction with #1209, but #12091 needs to be updated again. Can you update it  @cschanaj ? then we can merge this.

Also, docker on travis is failing for some reason.

cschanaj commented:
@cowlicks I have rebased this and merged #12091 here. I will update #11560 once this is merged. thanks!

cowlicks commented:
@cschanaj great work. Thank you.
